components/api/src/main/java/com/hotels/styx/api/extension/service/BackendService.kt

```diff
@@ -0,0 +1,280 @@
+  http://www.apache.org/licenses/LICENSE-2.0
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+package com.hotels.styx.api.extension.service
+
+import com.hotels.styx.api.Id
+import com.hotels.styx.api.Identifiable
+import com.hotels.styx.api.extension.Origin
+import com.hotels.styx.api.extension.Origin.checkThatOriginsAreDistinct
+import java.lang.IllegalArgumentException
+import java.lang.StringBuilder
+import java.net.URI
+import java.util.Objects
+import java.util.Optional
+
+/**
+ * Represents the configuration of an application (i.e. a backend service) that Styx can proxy to.
+ */
+class BackendService(
+    private val id: Id,
+    private val path: String,
+    private val connectionPoolSettings: ConnectionPoolSettings,
+    private val origins: Set<Origin>,
+    private val healthCheckConfig: HealthCheckConfig,
+    private val stickySessionConfig: StickySessionConfig,
+    private val rewrites: List<RewriteConfig>,
+    private val overrideHostHeader: Boolean,
+    private val responseTimeoutMillis: Int,
+    private val maxHeaderSize: Int,
+    private val tlsSettings: TlsSettings
+) : Identifiable {
+    /**
+     * A protocol used for the backend service. This can be either HTTP or HTTPS.
+     */
+    enum class Protocol {
+        HTTP, HTTPS
+    }
+
+    init {
+        checkThatOriginsAreDistinct(origins)
+        if (responseTimeoutMillis < 0) {
+            throw IllegalArgumentException("Request timeout must be greater than or equal to zero")
+        }
+    }
+
+    private constructor(builder: Builder): this(
+        id = builder.id!!,
+        path = builder.path!!,
+        connectionPoolSettings = builder.connectionPoolSettings!!,
+        origins = builder.origins!!,
+        healthCheckConfig = builder.healthCheckConfig!!,
+        stickySessionConfig = builder.stickySessionConfig!!,
+        rewrites = builder.rewrites!!,
+        overrideHostHeader = builder.overrideHostHeader!!,
+        responseTimeoutMillis = builder.responseTimeoutMillis!!,
+        maxHeaderSize = builder.maxHeaderSize!!,
+        tlsSettings = builder.tlsSettings!!
+    )
+
+    /**
+     * A builder for [BackendService].
+     */
+    class Builder {
+        var id: Id? = Id.GENERIC_APP
+            private set
+        var path: String? = "/"
+            private set
+        var origins: Set<Origin>? = emptySet()
+            private set
+        var connectionPoolSettings: ConnectionPoolSettings? = ConnectionPoolSettings.defaultConnectionPoolSettings()
+            private set
+        var stickySessionConfig: StickySessionConfig? = StickySessionConfig.stickySessionDisabled()
+            private set
+        var healthCheckConfig: HealthCheckConfig? = null
```

Comment:
If HealthCheckConfig isn't nullable on BackendService, this is sort of a problem if they don't set this, right? The solution might be to make this nullable on BackendService.